| Mini Group Assignment 4.2| Diffie Hellman Key Exchange |
| ----------- | ----------- |
|USD - CYBR504| March 2023|

|4.1.1|
|-----|
| Hands-on DH |

In [9]:
"""Function to calculate the shared private key (K)"""

# diffie hellman key exchange
def diffie_hellman_key_exchange(p, g, a, b):
    # pow(base, exp, mod=None)
    A = pow(g, a, p) # A = g^a mod p
    B = pow(g, b, p) # B = g^b mod p

    A_K = pow(B, a, p) # A_K = B^a mod p
    B_K = pow(A, b, p) # B_K = A^b mod p

    if A_K == B_K:
        print(A_K)
        print(B_K)
        print()
        print(f"Shared private key (K) = {A_K}")
        return A_K
    else:
        print('ERROR: Shared key not matching')
        return None

In [3]:
# TEAM SUBGROUP 1
# INPUTS
p = 10937 # prime number
g = 7 # primitive root
a = 4 # Matt's secret integer
b = 5 # Shashwati's secret integer

K = diffie_hellman_key_exchange(p, g, a, b)

371
371

Shared private key (K) = 371


What value did each of you get for K?
* 371

What were your p and q?
* p = 10937
* q = 7

What were your secret integers?
* 4 & 5

In [4]:
# TEAM SUBGROUP 2
# INPUTS
p = 10139 # prime number
g = 8 # primitive root
a = 5 # Connor's secret integer
b = 3 # Jerry's secret integer

K = diffie_hellman_key_exchange(p, g, a, b)

2981
2981

Shared private key (K) = 2981


What value did each of you get for K?
* 2981

What were your p and q?
* p = 10139
* q = 8

What were your secret integers?
* 5 & 3

| 4.1.2 |
|-------|

Assume Alice and Bob agreed on p=53, g=8.
Alice and Bob select secret integers, a and b.
Alice sent the value 48 to Bob.
Bob sent the value 35 to Alice.

In [8]:
# Alice & Bob
# INPUTS
p = 53 # prime number
g = 8 # primitive root
a = 48 # Alice's secret integer
b = 35 # Bob's secret integer

K = diffie_hellman_key_exchange(p, g, a, b)

10
10

Shared private key (K) = 10


Could you find K? What was it?
* 10

You would need to solve: (35^a mod 53 = 48^b mod 53)
How did you get your answer?
* We developed a function that calculates the shared secret key by implementing Diffie-Hellman's key exchange protocol.